cirq-google/cirq_google/devices/serializable_device.py

```diff
@@ -107,7 +107,7 @@ def __init__(
                 *[
                     g
                     for g in gate_definitions.keys()
-                    if isinstance(g, (cirq.Gate, type(cirq.Gate)))
+                    if isinstance(g, cirq.Gate) or issubclass(g, cirq.Gate)
```

Comment:
According to the type annotation on line 83 above, `g` is a *class object* for a class deriving from `cirq.Gate` (e.g. `cirq.XPowGate`). Hence, it should never be an *instance* of `cirq.Gate` (e.g. `cirq.X`). Consequently, `isinstance(g, cirq.Gate)` is always false. Also by the same type annotation, `g` represents a class derived from `cirq.Gate`, so `issubclass(g, cirq.Gate)` is always true. It seems to me that we should be able to remove this line altogether. Am I missing something?

Reply:
I tried to remove the type filter, but that causes ValueError on `import cirq_google` which trips on the definition of Sycamore device - https://github.com/quantumlib/Cirq/blob/b6d6577be61d216ce2f29f8c64ae5879cf3087d5/cirq-google/cirq_google/devices/known_devices.py#L315-L317

Somehow the build-up of the `gates_by_type` dictionary (used for the `gate_definitions` argument) adds a key with the FrozenCircuit type, which causes ValueError in Gateset initialization
https://github.com/quantumlib/Cirq/blob/b6d6577be61d216ce2f29f8c64ae5879cf3087d5/cirq-google/cirq_google/devices/serializable_device.py#L183-L197

The `issubclass(g, cirq.Gate)` removes the FrozenCircuit type, all other keys are of cirq.Gate type, at least for the Sycamore device.  I am not familiar with device deserialization details - perhaps the type check should be moved to `SerializableDevice.from_proto`?

The `isinstance(g, cirq.Gate)` branch in the filter does not affect the tests.

@MichaelBroughton - can you please check this logic?
The type filter was initially added in [12ebe83d](https://github.com/quantumlib/Cirq/commit/12ebe83d73385665ea57cb0e6886654d95381a7c#diff-010dde620095551aa3dbb209351a1d2c27f3ebc1cf9a4d11198da81561152246).
Do we need to check for `isinstance(g, cirq.Gate)`?

Reply:
Might this be related to subcircuits? @95-martin-orion 

I wonder whether we should even be filtering the FrozenCircuit instances out in the first place. Also, it'd be great to have a comment explaining why we do this and a type annotation that reflects the possibility of FrozenCircuits showing up as keys.

Reply:
OK. To make sure I understand the details: if a gate is a subcircuit then the key will be the *class object* `cirq.FrozenCircuit` or an *instance* of `cirq.FrozenCircuit`?

I think what this PR should do is:
 * update the type annotation on line 83 (and maybe other places), 
 * keep the filter,
 * simplify the condition(*),
 * add a comment about the reason for the filter.

---
(*) Details of this depend on the answer to question above. For example if the key is a class object then we can just remove the `isinstance` check. Otherwise, I suggest `if not isinstance(g, cirq.FrozenCircuit)` since in this case we can't use `issubclass` because IIRC it would raise a `TypeError` upon receiving something which is not a class object.

Reply:
We only define one (de)serializer for `CircuitOperation`, which declares its `internal_type` as the class object `cirq.FrozenCircuit`. The key in this PR will always be the class object.

Reply:
Thanks, @95-martin-orion !

@pavoljuhas Hopefully, the above clarifies the next steps to get this PR approved (sorry that a simple type check fix blew up like that). Also, while you're at it, it's probably worth it to try to answer the question why mypy does not report type errors despite the fact that our type annotation is too narrow. Is it perhaps related to the fact that the subcircuit case shows up in cirq_google?

Reply:
Done.  I think mypy got confused, because `SerializableGateSet.supported_internal_types()` returned
tuple without any typing for its content; that tuple was actually the source for `gate_definitions` keys.

https://github.com/quantumlib/Cirq/blob/b6d6577be61d216ce2f29f8c64ae5879cf3087d5/cirq-google/cirq_google/serialization/serializable_gate_set.py#L80-L81